### PR TITLE
Migrate to MBrace 0.10

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
This PR migrates the code samples to make use of the latest MBrace.Core/Azure version 0.10. These upgrades incorporate many changes to the core api, so many of the demos are no longer functional and/or irrelevant. 

This is the first release to incorporate the common runtime foundations of MBrace.Core, so code samples can now be used interchangeably between MBrace.Azure and MBrace.Thespian.

I will be pushing more commits that improve and update the samples in the next few days, merging to the master branch as soon as this becomes the default version used in brisk.

As always, please feel free to review the changes.

### CODE REVIEW

**Comment by eiriktsarpalis** [comment]
@dsyme @isaacabraham @palladin branch is ready for merge, please review.

File: HandsOnTutorial/1-hello-world.fsx [hunk 1]
```diff
@@ -1,15 +1,19 @@
 ﻿(*** hide ***)
-#load "credentials.fsx"
-
+#load "Thespian.fsx"
+#load "Azure.fsx"
```
Comment (dsyme) [inline HandsOnTutorial/1-hello-world.fsx]: Change the file names to ``AzureCluster.fsx`` and ``ThespianCluster.fsx``

Can you make it that to switch between the two you comment out one?

File: HandsOnTutorial/1-hello-world.fsx [hunk 2]
```diff
@@ -1,15 +1,19 @@
...
 open System
 open System.IO
 open MBrace.Core
-open MBrace.Azure
 open MBrace.Flow
 
+// Initialize client object to an MBrace cluster:
+let cluster = 
+//    getAzureClient() // comment out to use an MBrace.Azure cluster; don't forget to set the proper connection strings in Azure.fsx
+    initThespianCluster(4) // use a local cluster based on MBrace.Thespian; configuration can be adjusted using Thespian.fsx
```
Comment (dsyme) [inline HandsOnTutorial/1-hello-world.fsx]: This looks awkward.  Too many names for the first line of MBrace code :) "Azure", "Client", "Thespian", "init", "Cluster".  

If you switch to making the choice depdnent on which .fsx is used, then you can just stick to ``let cluster = GetCluster(config)``  where the script defines ``GetCluster``

File: HandsOnTutorial/1-hello-world.fsx [hunk 3]
```diff
@@ -17,47 +21,43 @@ common with MBrace.
...
 
-// You can view the history of processes:
-cluster.ShowProcessInfo()
-
 (** Now execute your first cloud workflow and get a handle to the running job: *)
-let job = 
+let task = 
     cloud { return "Hello world!" } 
-    |> cluster.CreateProcess
+    |> cluster.CreateCloudTask
```
Comment (dsyme) [inline HandsOnTutorial/1-hello-world.fsx]: I still find the ``CloudTask`` terminology difficult.  It raises do many questions and expectations in the user's mind ("how does this relate to .NET tasks?"), where ``Process`` and ``job`` made them feel happy (even with the overlap with Unix processes).

File: HandsOnTutorial/1-hello-world.fsx [hunk 4]
```diff
@@ -17,47 +21,43 @@ common with MBrace.
...
 
 // Block until the result is computed by the cluster
-let text = job.Result
+let text = task.Result
 
 (** Alternatively we can do this all in one line: *)
 let quickText = 
     cloud { return "Hello world!" } 
-    |> cluster.Run
+    |> cluster.RunOnCloud
```
Comment (dsyme) [inline HandsOnTutorial/1-hello-world.fsx]: Why not ``Run``.  A cloud task gets run in the cluster - that's not surprising.  RunOnCloud just poses more questions which a beginner shouldn't need to ask ("is the cloud different to the cluster", "why can I run anywhere else but the cloud"...)

File: HandsOnTutorial/1-hello-world.fsx [hunk 5]
```diff
@@ -67,26 +67,4 @@ In case you run into trouble, this can be used to clear all process
...
 
-cluster.ClearAllProcesses()
-
-(**
-If you need to get really heavy, you can reset the cluster, which clears 
-all process state in queues and storage. Other storage is left unchanged.
-Your worker roles may need to be manually rebooted (e.g. from the Azure 
-management console).
-*)
-// cluster.Reset()
```
Comment (dsyme) [inline HandsOnTutorial/1-hello-world.fsx]: What happened to ``cluster.Reset``, ``cluster.ShowSystemLogs``?

File: HandsOnTutorial/1-hello-world.fsx [hunk 6]
```diff
@@ -67,26 +67,4 @@ In case you run into trouble, this can be used to clear all process
...
-
-(** You can add your local machine to be a worker in the cluster. *)
-
-// MBraceAzure.SpawnLocal(config, 4)
-
-(** You can optionally look at worker (not client) logs for the last 5 minutes. *)
-
-// cluster.ShowSystemLogs(300.0)
-
-(** In this tutorial, you've run your very first work on an MBrace cluster
```
Comment (dsyme) [inline HandsOnTutorial/1-hello-world.fsx]: Please reinstate all the terminating comments that wrap up the tutorial.  These are the "summary" sections at the end of each page, e.g. http://www.m-brace.net/azure/HandsOnTutorial/1-hello-world.html.  They can be improved, but a concluding summary wrapping up what has been learnt is needed, so please make sure they are there.

File: HandsOnTutorial/10-using-cloud-dictionary.fsx [hunk 7]
```diff
@@ -1,13 +1,18 @@
...
 open MBrace.Core
-open MBrace.Store
 open MBrace.Azure
 open MBrace.Flow
 
+// Initialize client object to an MBrace cluster:
+let cluster = 
+//    getAzureClient() // comment out to use an MBrace.Azure cluster; don't forget to set the proper connection strings in Azure.fsx
+    initThespianCluster(4) // use a local cluster based on MBrace.Thespian; configuration can be adjusted using Thespian.fsx
+
```
Comment (dsyme) [inline HandsOnTutorial/10-using-cloud-dictionary.fsx]: Similar comments here

File: HandsOnTutorial/10-using-cloud-dictionary.fsx [hunk 8]
```diff
@@ -34,14 +36,13 @@ dict.TryFind "key-not-there" |> Async.RunSynchronously
 
 (** Next, perform contested, distributed updates: *) 
 let key = "contestedKey"
-let contestJob = 
+let contestTask = 
     [|1 .. 100|]
     |> CloudFlow.OfArray
-    |> CloudFlow.iterLocal(fun i -> dict.Transact(key, function None -> (),i | Some v -> (),i + v) |> Cloud.OfAsync)
-    |> cluster.CreateProcess
+    |> CloudFlow.iterLocal(fun i -> CloudDictionary.AddOrUpdate key (function None -> i | Some v -> i + v) dict |> Local.Ignore)
```
Comment (dsyme) [inline HandsOnTutorial/10-using-cloud-dictionary.fsx]: Why did you move away from instance notation?  It doesn't look right at all.

File: HandsOnTutorial/200-starting-a-web-server.fsx [hunk 9]
```diff
@@ -175,14 +175,10 @@ serverJob.ShowInfo()
...
 (** Use this to cancel the web server (via the distributed cancellationToken being cancelled): *)
 // serverJob.Cancel()
 
 (** Use this to wait for the web server to exit after cancellation: *)
-// serverJob.Resut
-
-
-
-
+// serverJob.Resut
```
Comment (dsyme) [inline HandsOnTutorial/200-starting-a-web-server.fsx]: There is a typo here (Resut --> Reset), it was there before though

File: HandsOnTutorial/300-distributed-multicore-parallel-search.fsx [hunk 10]
```diff
@@ -17,8 +22,8 @@ open MBrace.Flow
  Before running, edit credentials.fsx to enter your connection strings.
 **)
 
-(** First you connect to the cluster: *)
-let cluster = MBraceAzure.GetHandle(config)
+#load "lib/collections.fsx"
+#load "lib/mersenne.fsx"
```
Comment (dsyme) [inline HandsOnTutorial/300-distributed-multicore-parallel-search.fsx]: I recommend putting all ``#load`` at the top of the file.

File: HandsOnTutorial/400-using-cloud-channels.fsx [hunk 11]
```diff
@@ -0,0 +1,51 @@
+﻿(*** hide ***)
```
Comment (dsyme) [inline HandsOnTutorial/400-using-cloud-channels.fsx]: rename this file to ``using-cloud-queues.fsx``

File: HandsOnTutorial/400-using-cloud-channels.fsx [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+cloud workflows.
+ 
+Before running, edit credentials.fsx to enter your connection strings.
+*)
+
+(** Create an anonymous cloud channel: *) 
+let queue = CloudQueue.New<string>() |> cluster.RunOnCloud
+
+(** Send to the channel by scheduling a cloud process to do the send: *)
+CloudQueue.Enqueue (queue, "hello") |> cluster.RunOnCloud
```
Comment (dsyme) [inline HandsOnTutorial/400-using-cloud-channels.fsx]: Why don't you use instance notation?  Using static methods just doesn't give clean code for these kind of abstractions.

File: HandsOnTutorial/HandsOnTutorial.fsproj [hunk 13]
```diff
@@ -0,0 +1,787 @@
...
+    <None Include="999-misc-cpu-perf.fsx" />
+    <None Include="paket.references" />
+    <Content Include="libiomp5md.dll">
+      <Paket>True</Paket>
+    </Content>
+    <Content Include="MathNet.Numerics.MKL.dll">
+      <Paket>True</Paket>
+    </Content>
+    <Content Include="app.config" />
+    <Content Include="app.config" />
```
Comment (dsyme) [inline HandsOnTutorial/HandsOnTutorial.fsproj]: Two app.config for some reason

**Comment by dsyme** [comment]
This looks great.  Most my comments are about the API changes.  CloudTask still looks very awkward.   And I think choosing cloud provider is best done by the ``#load`` at the top?

File: HandsOnTutorial/1-hello-world.fsx (hunk 3 above)
Comment (eiriktsarpalis) [inline HandsOnTutorial/1-hello-world.fsx]: Could you create an issue in MBrace.Core for this? 

In short, the renaming was performed with a unification of the old Process<'T> (which lived in the client only and was non-serializable) with the serializable, MBrace.Core CloudTask<'T>, from which nested cloud tasks can be created. Also, we avoided the use the term 'job' since internally this is the name reserved for work items dequeued by machines in the cluster.

File: HandsOnTutorial/1-hello-world.fsx (hunk 4 above)
Comment (eiriktsarpalis) [inline HandsOnTutorial/1-hello-world.fsx]: There is a relevant discussion about this here: https://github.com/mbraceproject/MBrace.Core/pull/119

File: HandsOnTutorial/1-hello-world.fsx (hunk 5 above)
Comment (eiriktsarpalis) [inline HandsOnTutorial/1-hello-world.fsx]: The disappearance of cluster.Reset() is an omission that will be fixed. Logs can now be fetched on a per-task basis using `CloudTask.GetLogs`. System logs specific to the current process can be fetched by subscribing loggers to the cluster object using `cluster.AttachLogger`.

File: HandsOnTutorial/300-distributed-multicore-parallel-search.fsx (hunk 10 above)
Comment (eiriktsarpalis) [inline HandsOnTutorial/300-distributed-multicore-parallel-search.fsx]: The top declarations are omitted during document generation, so I thought it best to keep implementation-specific loads visible.

Comment (dsyme) [inline HandsOnTutorial/300-distributed-multicore-parallel-search.fsx]: That makes sense. I recommend they still be the first visible thing though

File: HandsOnTutorial/1-hello-world.fsx (hunk 2 above)
Comment (eiriktsarpalis) [inline HandsOnTutorial/1-hello-world.fsx]: Agreed, it is awkward indeed. I chose this as a lesser of many evils. My intention here was to avoid hiding any cluster initialization side-effects inside the scripts.

Comment (dsyme) [inline HandsOnTutorial/1-hello-world.fsx]: See https://github.com/mbraceproject/MBrace.StarterKit/pull/40
